src/serializer/ResidualHeapSerializer.js

```diff
@@ -1679,91 +1668,96 @@ export class ResidualHeapSerializer {
     return false;
   }
 
-  processAdditionalFunctionValues(): Map<FunctionValue, Array<BabelNodeStatement>> {
-    let rewrittenAdditionalFunctions: Map<FunctionValue, Array<BabelNodeStatement>> = new Map();
+  _serializeAdditionalFunctionGeneratorAndEffects(generator: Generator, postGeneratorCallback: () => void) {
+    let context = this._getContext();
+    return this._withGeneratorScope(generator, newBody => {
+      let oldCurBody = this.currentFunctionBody;
+      let oldSerialiedValueWithIdentifiers = this._serializedValueWithIdentifiers;
+      this.currentFunctionBody = newBody;
+      this._serializedValueWithIdentifiers = new Set(Array.from(this._serializedValueWithIdentifiers));
+      generator.serialize(context);
+      if (postGeneratorCallback) postGeneratorCallback();
+      this.currentFunctionBody = oldCurBody;
+      this._serializedValueWithIdentifiers = oldSerialiedValueWithIdentifiers;
+    });
+  }
+
+  // result -- serialize it, a return statement will be generated later
+  // Generator -- visit all entries
+  // Bindings -- only need to serialize bindings if they're captured by some nested function?
+  //          -- need to apply them and maybe need to revisit functions in ancestors to make sure
+  //          -- we don't overwrite anything they capture
+  // PropertyBindings -- visit any property bindings that aren't to createdobjects
+  // CreatedObjects -- should take care of itself
+  _serializeAdditionalFunctionEffects(additionalFunctionValue: FunctionValue, effects: Effects) {
+    let [result, , , modifiedProperties: Map<PropertyBinding, void | Descriptor>, createdObjects] = effects;
+    for (let propertyBinding of modifiedProperties.keys()) {
+      let binding: PropertyBinding = ((propertyBinding: any): PropertyBinding);
+      let object = binding.object;
+      if (object instanceof ObjectValue && createdObjects.has(object)) continue;
+      if (object.refuseSerialization) continue;
+      if (object.isIntrinsic()) continue;
+      invariant(object instanceof ObjectValue);
+      this._emitProperty(object, binding.key, binding.descriptor, true);
+    }
+    invariant(result instanceof Value);
```

Comment:
Is this because of a precondition? If so, update the function comment to make it clear that effects[0] must be a Value.